com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/JarSignatureVerifier.java

```diff
@@ -0,0 +1,209 @@
+        List<JarEntry> nonSignedJarEntries = getNonSignedJarEntries(nonMetaInformationJarEntries);
+        if (!nonSignedJarEntries.isEmpty()) {
+            if (nonMetaInformationJarEntries.size() == nonSignedJarEntries.size()) {
+                throw new SLException(Messages.THE_ARCHIVE_IS_NOT_SIGNED);
+            }
+            throw new SLException(Messages.THE_ARCHIVE_CONTAINS_UNSIGNED_FILES, getJarEntriesNames(nonSignedJarEntries));
+        }
+    }
+
+    private List<JarEntry> getNonMetaInformationJarEntries(List<JarEntry> jarEntries) {
+        return jarEntries.stream()
+                         .filter(jarEntry -> !isMetaInformation(jarEntry))
+                         .collect(Collectors.toList());
+    }
+
+    private boolean isMetaInformation(JarEntry jarEntry) {
+        return jarEntry.getName()
+                       .startsWith(META_INF);
+    }
+
+    private List<JarEntry> getNonSignedJarEntries(List<JarEntry> jarEntries) {
+        return jarEntries.stream()
+                         .filter(jarEntry -> !isSigned(jarEntry))
+                         .collect(Collectors.toList());
+    }
+
+    private boolean isSigned(JarEntry jarEntry) {
+        return !ArrayUtils.isEmpty(jarEntry.getCertificates());
+    }
+
+    private String getJarEntriesNames(List<JarEntry> jarEntries) {
+        return jarEntries.stream()
+                         .map(ZipEntry::getName)
+                         .collect(Collectors.joining(System.lineSeparator()));
+    }
+
+    private List<JarEntry> getSignedJarEntries(List<JarEntry> jarEntries) {
+        return jarEntries.stream()
+                         .filter(this::isSigned)
+                         .collect(Collectors.toList());
+    }
+
+    private List<X509Certificate> toX509Certificates(Certificate[] certificates) {
+        return Arrays.stream(certificates)
+                     .map(certificate -> (X509Certificate) certificate)
+                     .collect(Collectors.toList());
+    }
+
+    private void validateCertificateChain(List<X509Certificate> targetCertificates, List<X509Certificate> candidateCertificateChain,
+                                          String certificateCN) {
+        boolean certificateCNMatches = false;
+        boolean targetCertificatesMatchEntryCertificate = false;
+        for (X509Certificate certificate : candidateCertificateChain) {
+            checkValidityOfCertificate(certificate);
+            if (!certificateCNMatches) {
+                certificateCNMatches = certificateCNMatches(certificate, certificateCN);
+            }
+            if (!targetCertificatesMatchEntryCertificate) {
+                targetCertificatesMatchEntryCertificate = doTargetCertificatesMatchEntryCertificate(targetCertificates, certificate);
+            }
+        }
+        validateCertificateChain(certificateCNMatches, targetCertificatesMatchEntryCertificate, targetCertificates, certificateCN);
```

Comment:
I preferred it when this method was inlined. Yes, the method was long, but the reason why we extract methods is to improve readability. This only works if each method handles a separate concept/operation that is understandable on its own. If you need to read one method in order to understand what another does, then there's something wrong with those methods.

Have you tested the performance, BTW? If this isn't slow, then the whole for loop could be split into two separate loops. Yes, it would be 2 times slower, but it would be much easier to read and undertand.